File: Scripts/tracker.js
// = Name
// tracker.js
// Douglass - Cross-browser, cross-platform alternative to CTA BusTracker
//
// = Description
// The ChicagoTransit module queries the CTA BusTracker Service using the
// official version 1.0 API. CTA uses the BusTime system to provide developers
// with real-time route information through BusTracker. The BusTracker Service
// allows deveopers to make HTTP requests and recieve XML responses. The CTA
// provides {official API documentation (PDF)}[http://www.transitchicago.com/assets/1/developer_center/BusTime_Developer_API_Guide.pdf],
// and you can request an access key at the {API homepage}[http://www.transitchicago.com/developers/bustracker.aspx].
//
// Requires:: jQuery (1.4.2)
// Version::  01
// Updated::  2010-03-23
// Author::   Sean Clemmer


// == Tracker
// JavaScript implementation of the CTA BusTracker system.
// Most methods use a callback function, which is loaded with the results of the query.
// Response errors from the system load the callback with an error property.
// == Example Usage
// tracker = new Tracker();
function Tracker() {

  // === Usage
  // Create a Tracker error and pass it to the callback
  // === Parameters
  // - message: Human-readable error description
  // - callback: Function to pass error with
  var create_error_with = function( message, callback ) {
    var error = {
      'message' : message,
      'error'   : true
    };
    callback.call( this, error );
  };


  // === Usage
  // Handle CTA errors and pass them to the callback
  // === Parameters
  // - response: XML document processed in the $.ajax request
  // - callback: Function to pass error with
  var handle_error_with = function( response, callback ) {
    if ( $( response ).find( 'error' ).length != 0 ) {
      create_error_with( 'CTA Error: ' + $( response ).find( 'error > msg' ).text(), callback );
      return true;
    } else {
      return false;
    }
  };


  // === Usage
  // Convert an XML response into a JSON object
  // === Parameters
  // - response: XML document processed in the $.ajax request
  // - root: Name of an XML tag to look up properties
  // - properties: JSON dictionary of CTA tags and property equivalents
  var jsonify = function( response, root, properties ) {
    var items = new Array();
    $( response ).find( root ).each( function() {
      var item = {};
      var this_response = $( this );
      $.each( properties, function( cta_tag, property ) {
        if ( property == 'delayed' ) {
          // I wonder if there is a better way of handling this case...
          item[property] = this_response.find( cta_tag ).text() == '' ? false : true;
        } else {
          item[property] = this_response.find( cta_tag ).text();
        }
      });
      items.push( item );
    });
    return items;
  };


  // === Usage
  // Construct a new instance of the Tracker.
  // === Parameters
  // - api_key: 25-digit CTA BusTracker API access key (required)
  // === Example Usage
  // tracker.initialize( CTA_API_KEY );
  this.initialize = function( api_key ) {
    this.api_key   = api_key;
    this.api_url   = 'http://www.ctabustracker.com/bustime/api/v1/';
    this.key_query = '?key=' + this.api_key;
  };


  // === Usage
  // Retrieve the system date and (local) time.
  // === Base URL
  // http://www.ctabustracker.com/bustime/api/v1/gettime
  // === Example Usage
  // tracker.time( function( time ) {
  //   alert( time );
  // });
  this.time = function( callback ) {
    var request = this.api_url + 'gettime' + this.key_query;

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var time = $( response ).find( 'tm' ).text();
          callback.call( this, time);
        }
      }
    });
  };


  // === Usage
  // Retrieve information for vehicles serviced by the specified list or routes.
  // === Base URL
  // http://www.ctabustracker.com/bustime/api/v1/getvehicles
  // === Parameters
  // - routes: List of CTA route designators (not available with vehicles parameter)
  // - vehicles: List of CTA vehicle IDs (not available with routes parameter)
  // At least one required. If both routes and vehicles are provided, routes are used.
  // === Example Usage
  // tracker.vehicles( { 'routes': '55' },
  //   function( vehicles ) {
  //     if ( vehicles.error == true ) {
  //       alert( vehicles.message );
  //     } else {
  //       $.each( vehicles, function( index, vehicle ) {
  //         alert( vehicle.number );
  //       });
  //     }
  //   }
  // );
  this.vehicles = function( params, callback ) {
    var request = this.api_url + 'getvehicles' + this.key_query;
    
    if ( !params.routes && !params.vehicles ) {
      create_error_with( "No 'routes' or 'vehicles' provided", callback );
      return;
    }

    params.routes ? request += '&rt=' + params.routes : undefined;
    params.vehicles ? request += '&vidd=' + params.vehicles : undefined;

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var vehicles = jsonify( response, 'vehicle', {
            'vid'    : 'number',
            'tmstmp' : 'timestamp',
            'lat'    : 'latitude',
            'lon'    : 'longitude',
            'hdg'    : 'heading',
            'pid'    : 'pattern',
            'pdist'  : 'distance',
            'rt'     : 'route',
            'des'    : 'destination',
            'dly'    : 'delayed'
          });
          callback.call( this, vehicles );
        }
      }
    });
  };


  // === Usage
  // Retrieve a list of routes serviced by the system.
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getroutes
  // === Example Usage
  // tracker.routes(
  //   function( routes ) {
  //     if ( routes.error == true ) {
  //       alert( routes.message );
  //     } else {
  //       $.each( routes, function( index, route ) {
  //         alert( route.number );
  //       });
  //     }
  //   }
  // );
  this.routes = function( callback ) {
    var request = this.api_url + 'getroutes' + this.key_query;

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var routes = jsonify( response, 'route', {
            'rt'   : 'number',
            'rtnm' : 'description'
          });
          callback.call( this, routes );
        }
      }
    });
  };


  // === Usage
  // Retrieve a set of directions serviced by the specified route.
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getdirections
  // === Parameters
  // - route: Single route designator (required)
  // === Example usage
  // tracker.directions( { 'route': '55' },
  //  function( directions ) {
  //     if ( directions.error == true ) {
  //       alert( directions.message );
  //     } else {
  //       $.each( directions, function( index, direction ) {
  //         alert( direction.value );
  //       });
  //     }
  //   }
  // );
  this.directions = function( params, callback ) {
    var request = this.api_url + 'getdirections' + this.key_query;

    if ( params.route ) {
      request += '&rt=' + params.route;
    } else {
      create_error_with( "No 'route' provided", callback );
      return;
    }

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          // This would be nice to abstact, like the others...
          var directions = new Array();
          $( response ).find( 'dir' ).each( function() {
            var direction = {
              'value' : $( this ).text()
            };
            directions.push( direction );
          });
          callback.call( this, directions );
        }
      }
    });
  };


  // === Usage
  // Retrieve a set of stoped serviced by the specified route and direction.
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getstops
  // === Parameters
  // - route: Single route designator (required)
  // - direction: Sing route direction (required)
  // == Example Usage
  // tracker.stops( { 'route': '55', 'direction': 'East bound' },
  //   function( stops ) {
  //     if ( stops.error == true ) {
  //       alert( stops.message );
  //     } else {
  //       $.each( stops, function( index, stop ) {
  //         alert( stop.number );
  //       });
  //     }
  //   }
  // );
  this.stops = function( params, callback ) {
    var request = this.api_url + 'getstops' + this.key_query;
    
    if ( params.route ) {
      if ( params.direction ) {
        request += '&rt=' + params.route + '&dir=' + params.direction;
      } else {
        create_error_with( "No 'direciton' provided", callback );
        return;
      }
    } else {
      create_error_with( "No 'route' provided", callback );
      return;
    }

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var stops = jsonify( response, 'stop', {
              'stpid' : 'number',
              'stpnm' : 'description',
              'lat'   : 'latitidue',
              'lon'   : 'longitude'
          });
          callback.call( this, stops );
        }
      }
    });
  };


  // === Usage
  // Retrieve a set of patterns serviced by the specified list or routes.
  // Patterns are composed of geo-positional points that may be used to construct a map of the pattern (i.e. route variation).
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getpatterns
  // === Parameters
  // - route: Single route designator (not available with patterns parameter)
  // - patterns: List of pattern IDs (not available with route parameter)
  // At least one required. If both route and patterns are provided, route is used.
  // === Example Usage
  // tracker.patterns( { 'route': '55' },
  //   function( patterns ) {
  //     if ( patterns.error == true ) {
  //       alert( patterns.message );
  //     } else {
  //       $.each( patterns, function( index, pattern ) {
  //         alert( pattern.number );
  //       });
  //     }
  //   }
  // );
  this.patterns = function( params, callback ) {
    var request = this.api_url + 'getpatterns' + this.key_query;

    if ( params.route || params.patterns ) {
      params.route   ? request += '&rt='  + params.route   : undefined;
      params.pattern ? request += '&pid=' + params.pattern : undefined;
    } else {
      create_error_with( "No 'route' or 'patterns' provided", callback );
      return;
    }

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var patterns = jsonify( response, 'ptr', {
            'pid'   : 'number',
            'ln'    : 'length',
            'rtdir' : 'direction',
            'pt'    : 'points'
          });
          callback.call( this, patterns );
        }
      }
    });
  };

  // === Usage
  // Retrieve a set of predictions for the specified stops or vehicles.
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getpredictions
  // === Parameters
  // - stops: List of stop IDs (not available with vehicles parameter)
  // - routes: List of route designators (optional, only available with stops parameter)
  // - vehicles: List of vehicle IDs (not available with stops parameter)
  // - ceiling: Maximum number of predictions to retrieve (optional)
  // At least stops or vehicles required. If both stops and vehiles are provided, stops are used.
  // === Example Usage
  // tracker.predictions( { 'vehicles': '1105' },
  //   function( predictions ) {
  //     if ( predictions.error == true ) {
  //       alert( predictions.message );
  //     } else {
  //       $.each( predictions, function( index, prediction ) {
  //         alert( prediction.route );
  //       });
  //     }
  //   }
  // );
  this.predictions = function( params, callback ) {
    var request = this.api_url + 'getpredictions' + this.key_query;

    if ( !params.stops && !params.vehicles ) {
      create_error_with( "No 'stops' or 'vehicles' provided", callback );
      return;
    } else {
      params.stops    ? request += '&stpid=' + params.stops    : undefined;
      params.routes   ? request += '&rt='    + params.routes   : undefined;
      params.vehicles ? request += '&vid='   + params.vehicles : undefined;
      params.ceiling  ? request += '&top='   + params.ceiling  : undefined;
    }

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var predictions = jsonify( response, 'prd', {
            'tmstmp' : 'timestamp',
            'typ'    : 'type',
            'stpid'  : 'stop',
            'stpnm'  : 'description',
            'vid'    : 'vehicle',
            'dstp'   : 'distance',
            'rt'     : 'route',
            'des'    : 'destination',
            'prdtm'  : 'prediction',
            'dly'    : 'delayed'
          });
          callback.call( this, predictions );
        }
      }
    });
  };


  // === Usage
  // Retrieve a set of service bulletins in effect for the specified routes or stops.
  // === Base Url
  // http://www.ctabustracker.com/bustime/api/v1/getservicebulletins
  // === Parameters
  // - routes: List of route designators (required without stops parameter)
  // - direction: Single route direction (optional)
  // - stops: List of stop IDs (required without routes parameter)
  // === Example Usage
  // tracker.bulletins( { 'routes': '55' },
  //   function( bulletins ) {
  //     if ( bulletins.error == true ) {
  //       alert( bulletins.message );
  //     } else {
  //       $.each( bulletins, function( index, bulletin ) {
  //         alert( bulletin.number );
  //       });
  //     }
  //   }
  // );
  this.bulletins = function( params, callback ) {
    var request = this.api_url + 'getservicebulletins' + this.key_query;

    if ( !params.stops && !params.routes ) {
      create_error_with( "No 'routes' or 'stops' provided", callback );
      return;
    } else {
      params.direction ? request += '&dir='   + params.direction : undefined;
      params.routes    ? request += '&rt='    + params.routes    : undefined;
      params.stops     ? request += '&stpid=' + params.stops     : undefined;
    }

    $.ajax({
      url     : request,
      success : function( response ) {
        if ( !handle_error_with( response, callback ) ) {
          var bulletins = jsonify( response, 'sb', {
            'nm'   : 'description',
            'sbj'  : 'subject',
            'dtl'  : 'detail',
            'brf'  : 'brief',
            'prty' : 'priority',
            'srvc' : 'service'
          });
          callback.call( this, bulletins );
        }
      }
    });
  };
}

File: Scripts/Tracker.rb
# = Name
# Chicago Transit Authority BusTracker Service API
# 
# = Description
# The ChicagoTransit module queries the CTA BusTracker Service using the
# official version 1.0 API. CTA uses the BusTime system to provide developers
# with real-time route information through BusTracker. The BusTracker Service
# allows deveopers to make HTTP requests and recieve XML responses. The CTA
# provides {official API documentation (PDF)}[http://www.transitchicago.com/assets/1/developer_center/BusTime_Developer_API_Guide.pdf],
# and you can request an access key at the {API homepage}[http://www.transitchicago.com/developers/bustracker.aspx].
# A {wiki}[http://chicagowiki.transitapi.com/Home] was also created for the API.
#
# = Authors
# Sean Clemmer
module ChicagoTransit
  require 'nokogiri'
  require 'open-uri'
  require 'cgi'

  # === Usage
  # Convert a CTA-formatted timestamp into the standard Ruby format (without seconds).
  # === Parameters
  # - cta_time: CTA-formatted timestamp (required)
  def make_ruby_time_without_seconds_from( cta_time )
    time = cta_time.split
    t =  time[1].split(/:/).reverse
    t << time[0].gsub(/^....../, '')
    t << time[0].gsub(/^..../, '').gsub(/..$/, '')
    t << time[0].gsub(/....$/, '')
    Time.mktime( t[4], t[3], t[2], t[1], t[0] )
  end

  # === Usage
  # Convert a CTA-formatted timestamp into the standard Ruby format.
  # === Parameters
  # - cta_time: CTA-formatted timestamp (required)
  def make_ruby_time_from( cta_time )
    time = cta_time.split
    t =  time[1].split(/:/).reverse
    t << time[0].gsub(/^....../, '')
    t << time[0].gsub(/^..../, '').gsub(/..$/, '')
    t << time[0].gsub(/....$/, '')
    Time.mktime( t[5], t[4], t[3], t[2], t[1], t[0] )
  end

  # === Usage
  # Convert a standard Ruby timestamp into the CTA format (without seconds).
  # === Parameters
  # - time: Standard Ruby timestamp (optional)
  def make_cta_time_without_seconds_from( time=Time.now )
    return time.strftime( "%Y%m%d %H:%M" )
  end

  # === Usage
  # Convert a standard Ruby timestamp into the CTA format.
  # === Parameters
  # - time: Standard Ruby timestamp (optional)
  def make_cta_time_from( time=Time.now )
    return time.strftime( "%Y%m%d %H:%M:%S" )
  end

  # = Tracker
  # Ruby implementation of the CTA BusTracker system.
  # Most instance methods implemented here return an array or an array of hashes when successful.
  # Response errors from the system raise a runtime error. Other errors may be resuced to nil.
  # = Example Usage
  # Afer requiring this file and including the ChicagoTransit module in +irb+:
  #  >> tracker = ChicagoTransit::Tracker.new( 'your_api_key_here' )
  #  => #<ChicagoTransit::Tracker:0x1006197d8>
  #  >> tracker.system_time
  #  => Mon Jan 18 20:06:04 -0600 2010
  #  >> tracker.routes                      
  #  => [{"number"=>"1", "description"=>"Indiana/Hyde Park"}, ...
  #  >> tracker.vehicles :routes => [ '55' ]       
  #  => [{"number"=>"1731", "route"=>"55", "latitude"=>"41.791778564453125", ...
  #  >> tracker.bulletins :routes => [ '171' ]
  #  => [{"service"=>#<Nokogiri::XML::Element:0x80dccec4 name="srvc" ...
  class Tracker
    attr_reader :api_key

    # === Usage
    # Construct a new instance of the Tracker.
    # === Parameters
    # - API key: 25-digit CTA BusTracker API access key (required)
    def initialize( api_key )
      @@api_key   = api_key
      @@api_url   = 'http://www.ctabustracker.com/bustime/api/v1/'
      @@key_query = '?key=' + @@api_key
    end

    # === Usage
    # Retrieve the system date and (local) time.
    # Uses ChicagoTransit::make_ruby_time_from( cta_time ) to convert the CTA format to Ruby standards.
    # === Base URL
    # http://www.ctabustracker.com/bustime/api/v1/gettime
    def system_time
      request  = @@api_url + 'gettime' + @@key_query
      response = Nokogiri::XML( open( request ) )
      year = '', month = '', day = '', hour = '', min = '', sec = ''
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      return ChicagoTransit::make_ruby_time_from response.xpath( '//bustime-response/tm' ).first.content rescue nil
    end

    # === Usage
    # Retrieve information for vehicles serviced by the specified list or routes.
    # === Base URL
    # http://www.ctabustracker.com/bustime/api/v1/getvehicles
    # === Parameters
    # - routes: List of CTA route designators (not available with vehicles parameter)
    # - vehicles: List of CTA vehicle IDs (not available with routes parameter)
    # At least one required. If both routes and vehicles are provided, routes are used.
    def vehicles( params )
      request    = @@api_url + 'getvehicles' + @@key_query
      params[:routes].map! { |i| CGI::escape i } rescue nil
      params[:vehicles].map! { |i| CGI::escape i } rescue nil
      routes     = params[:routes].join( ',' ) rescue ''
      vehicles   = params[:vehicles].join( ',' ) rescue ''
      response   = Nokogiri::XML( open( request + '&rt=' + routes + '&vidd=' + vehicles ) )
      vehicles   = Array.new
      properties = {
        'number'      => 'vid',
        'timestamp'   => 'tmstmp',
        'latitude'    => 'lat',
        'longitude'   => 'lon',
        'heading'     => 'hdg',
        'pattern'     => 'pid',
        'distance'    => 'pdist',
        'route'       => 'rt',
        'destination' => 'des',
        'delayed'     => 'dly'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/vehicle/vid').length.times do |index|
        vehicle = Hash.new
        properties.each do |name, rep|
          unless name == 'timestamp'
            vehicle["#{name}"] = response.xpath("//bustime-response/vehicle[#{index+1}]/#{rep}").first.content rescue nil
          else
            vehicle["#{name}"] = ChicagoTransit::make_ruby_time_without_seconds_from response.xpath("//bustime-response/vehicle[#{index+1}]/#{rep}").first.content rescue nil
          end
        end
        vehicles << vehicle
      end
      return vehicles
    end

    # === Usage
    # Retrieve a list of routes serviced by the system.
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getroutes
    def routes
      request    = @@api_url + 'getroutes' + @@key_query
      response   = Nokogiri::XML( open( request ) )
      routes     = Array.new
      properties = {
        :number      => 'rt',
        :description => 'rtnm'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/route/rt').length.times do |index|
        route = Hash.new
        properties.each do |name, rep|
          route["#{name}"] = response.xpath("//bustime-response/route[#{index+1}]/#{rep}").first.content rescue nil
        end
        routes << route
      end
      return routes
    end

    # === Usage
    # Retrieve a set of directions serviced by the specified route.
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getdirections
    # === Parameters
    # - route: Single route designator (required)
    def directions( params )
      request    = @@api_url + 'getdirections' + @@key_query
      route      = CGI::escape params[:route].to_s
      response   = Nokogiri::XML( open( request + '&rt=' + route ) )
      directions = Array.new
      properties = {
        'direction' => 'dir'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/dir').length.times do |index|
        direction = Hash.new
        properties.each do |name, rep|
          direction["#{name}"] = response.xpath("//bustime-response/#{rep}[#{index+1}]").first.content rescue nil
        end
        directions << direction
      end
      return directions
    end

    # === Usage
    # Retrieve a set of stoped serviced by the specified route and direction.
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getstops
    # === Parameters
    # - route: Single route designator (required)
    # - direction: Sing route direction (required)
    def stops( params )
      request    = @@api_url + 'getstops' + @@key_query
      route      = CGI::escape params[:route].to_s
      direction  = CGI::escape params[:direction].to_s
      response   = Nokogiri::XML( open( request + '&rt=' + route + '&dir=' + direction ) )
      stops      = Array.new
      properties = {
        'number'      => 'stpid',
        'description' => 'stpnm',
        'latitude'    => 'lat',
        'longitude'   => 'lon'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/stop').length.times do |index|
        stop = Hash.new
        properties.each do |name, rep|
          stop["#{name}"] = response.xpath("//bustime-response/stop[#{index+1}]/#{rep}").first.content rescue nil
        end
        stops << stop
      end
      return stops
    end

    # === Usage
    # Retrieve a set of patterns serviced by the specified list or routes.
    # Patterns are composed of geo-positional points that may be used to construct a map of the pattern (i.e. route variation).
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getpatterns
    # === Parameters
    # - route: Single route designator (not available with patterns parameter)
    # - patterns: List of pattern IDs (not available with route parameter)
    # At least one required. If both route and patterns are provided, route is used.
    def patterns( params )
      # Todo: Encode points into their own hash  
      request    = @@api_url + 'getpatterns' + @@key_query
      params[:patterns].map! { |i| CGI::escape i } rescue nil
      route      = CGI::escape params[:route].to_s
      numbers    = params[:patterns].join(',') rescue ''
      response   = Nokogiri::XML( open( request + '&rt=' + route + '&pid=' + numbers ) )
      patterns   = Array.new
      properties = {
        'number'      => 'pid',
        'length'      => 'ln',
        'direction'   => 'rtdir',
        'points'      => 'pt'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/ptr').length.times do |index|
        pattern = Hash.new
        properties.each do |name, rep|
          unless name == 'points'
            pattern["#{name}"] = response.xpath("//bustime-response/ptr[#{index+1}]/#{rep}").first.content rescue nil
          else
            pattern["#{name}"] = response.xpath("//bustime-response/ptr[#{index+1}]/#{rep}").first rescue nil
          end
        end
        patterns << pattern
      end
      return patterns
    end

    # === Usage
    # Retrieve a set of predictions for the specified stops or vehicles.
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getpredictions
    # === Parameters
    # - stops: List of stop IDs (not available with vehicles parameter)
    # - routes: List of route designators (optional, only available with stops parameter)
    # - vehicles: List of vehicle IDs (not available with stops parameter)
    # - ceiling: Maximum number of predictions to retrieve (optional)
    # At least stops or vehicles required. If both stops and vehiles are provided, stops are used.
    def predictions( params )
      request     = @@api_url + 'getpredictions' + @@key_query
      params[:stops].map! { |i| CGI::escape i } rescue nil
      params[:routes].map! { |i| CGI::escape i } rescue nil
      params[:vehicles].map! { |i| CGI::escape i } rescue nil
      stops       = params[:stops].join(',') rescue ''
      routes      = params[:routes].join(',') rescue ''
      vehicles    = params[:vehicles].join(',') rescue ''
      ceiling     = CGI::escape params[:ceiling].to_s
      response    = Nokogiri::XML( open( request + '&stpid=' + stops + '&rt=' + routes + '&vid=' + vehicles + '&top=' + ceiling ) )
      predictions = Array.new
      properties  = {
        'timestamp'   => 'tmstmp',
        'type'        => 'typ',
        'stop'        => 'stpid',
        'description' => 'stpnm',
        'vehicle'     => 'vid',
        'distance'    => 'dstp',
        'route'       => 'rt',
        'destination' => 'des',
        'prediction'  => 'prdtm',
        'delayed'     => 'dly'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/prd').length.times do |index|
        prediction = Hash.new
        properties.each do |name, rep|
          unless name == 'prediction' or name == 'timestamp'
            prediction["#{name}"] = response.xpath("//bustime-response/prd[#{index+1}]/#{rep}").first.content rescue nil
          else
            prediction["#{name}"] = ChicagoTransit::make_ruby_time_without_seconds_from response.xpath("//bustime-response/prd[#{index+1}]/#{rep}").first.content rescue nil
          end
        end
        predictions << prediction
      end
      return predictions
    end

    # === Usage
    # Retrieve a set of service bulletins in effect for the specified routes or stops.
    # === Base Url
    # http://www.ctabustracker.com/bustime/api/v1/getservicebulletins
    # === Parameters
    # - routes: List of route designators (required without stops parameter)
    # - direction: Single route direction (optional)
    # - stops: List of stop IDs (required without routes parameter)
    def bulletins( params )
      # Todo: Encode services into their own hash
      request     = @@api_url + 'getservicebulletins' + @@key_query
      params[:routes].map! { |i| CGI::escape i } rescue nil
      params[:stops].map! { |i| CGI::escape i } rescue nil
      routes      = params[:routes].join(',') rescue ''
      direction   = CGI::escape params[:direction].to_s rescue ''
      stops       = params[:stops].join(',') rescue ''
      response    = Nokogiri::XML( open( request + '&stpid=' + stops + '&rt=' + routes + '&dir=' + direction ) )
      bulletins   = Array.new
      properties  = {
        'description' => 'nm',
        'subject'     => 'sbj',
        'detail'      => 'dtl',
        'brief'       => 'brf',
        'priority'    => 'prty',
        'service'     => 'srvc'
      }
      response.xpath('//bustime-response/error').length.times do |index|
        raise 'CTA: ' + response.xpath("//bustime-response/error[#{index+1}]/msg").first.content
      end
      response.xpath('//bustime-response/sb').length.times do |index|
        bulletin = Hash.new
        properties.each do |name, rep|
          unless name == 'service'
            bulletin["#{name}"] = response.xpath("//bustime-response/sb[#{index+1}]/#{rep}").first.content rescue nil
          else
            bulletin["#{name}"] = response.xpath("//bustime-response/sb[#{index+1}]/#{rep}").first rescue nil
          end
        end
        bulletins << bulletin
      end
      return bulletins
    end
  end
end